### PR TITLE
[spring-cloud] Refactor app create update and deploy logic

### PR DESCRIPTION
**Pull request by yuwzho**
---

This PR refactors the logic when app create, app deploy, app update and deployment create.

> This PR will NOT upgrade the version and will update together with other PRs.

App's command usually operates an Spring/Apps and the active Spring/Apps/Deployments under the app.
The general idea of these command is putting all input command in parameter dict and let the Resource factory to construct the payload.
 - _app_factory construct the App's properties
 - _deployment_factory construct the Deployment's properties, which includes source and other settings.
 - There are 5 types for different deployment source types, here use _deployment_source_factory to construct the payload according to the source_type
 - A deployment must consume a path can be deployable, it can be a relative path, custom container or build result resource id,
   - _deployment_deployable_factory determines the deployable type and upload necessary binary/code to the service when constructing the deployable_path.
   - _deployment_uploadable_factory will upload the local file to a given destination, or compress the local folder and upload according to the parameter.

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/spring-cloud/azext_spring_cloud/commands.py [hunk 1]
```diff
@@ -58,11 +63,14 @@ def load_command_table(self, _):
         g.custom_command('repo update', 'config_repo_update')
         g.custom_command('repo list', 'config_repo_list')
 
-    with self.command_group('spring-cloud app', client_factory=cf_spring_cloud_20220101preview,
+    with self.command_group('spring-cloud app', custom_command_type=app_command,
```
Comment (yuwzho) [inline src/spring-cloud/azext_spring_cloud/commands.py:66]: The 4 commands' traffic will be routing to the app.py
- `spring-cloud app create`
- `spring-cloud app update`
- `spring-cloud app deploy`
- `spring-cloud app deployment create`

**Comment by yonzhan** [comment]
spring-cloud

**Comment by yuwzho** [comment]
ready_to_review

File: src/spring-cloud/azext_spring_cloud/_buildservices_factory.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    def _get_upload_info(self):
+        try:
+            response = self.client.build_service.get_resource_upload_url(self.resource_group, self.service, self.name)
+            if not response.upload_url:
+                raise CLIError("Failed to get a SAS URL to upload context.")
+            return response
+        except CloudError as e:
+            raise CLIError("Failed to get a SAS URL to upload context. Error: {}".format(e.message))
+        except AttributeError as e:
+            raise CLIError("Failed to get a SAS URL to upload context. Error: {}".format(e))
```
Comment (zhoxing-ms) [inline src/spring-cloud/azext_spring_cloud/_buildservices_factory.py:58]: Please use the specific error type instead of `CLIError`

File: src/spring-cloud/azext_spring_cloud/_buildservices_factory.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            env={"BP_MAVEN_BUILT_MODULE": target_module} if target_module else None)
+        build = models.Build(properties=properties)
+        try:
+            return self.client.build_service.create_or_update_build(self.resource_group,
+                                                                    self.service,
+                                                                    self.name,
+                                                                    app,
+                                                                    build).properties.triggered_build_result.id
+        except (AttributeError, CloudError) as e:
+            raise CLIError("Failed to create or update a build. Error: {}".format(e.message))
```
Comment (zhoxing-ms) [inline src/spring-cloud/azext_spring_cloud/_buildservices_factory.py:76]: Same as above

File: src/spring-cloud/azext_spring_cloud/_buildservices_factory.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            sleep(5)
+            result = self._get_build_result(build_result_id)
+
+        if not build_log_streaming_available:
+            logger.warning("Cannot show real time build logs at this moment")
+            self._try_print_build_logs(build_result_id)
+
+        if result.properties.provisioning_state != "Succeeded":
+            log_url = self._try_get_build_log_url(build_result_id)
+            raise CLIError("Failed to build docker image, please check the build logs {} and retry.".format(log_url))
```
Comment (zhoxing-ms) [inline src/spring-cloud/azext_spring_cloud/_buildservices_factory.py:103]: Same as above

File: src/spring-cloud/azext_spring_cloud/app.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+def _ensure_app_not_exist(client, resource_group, service, name):
+    app = None
+    try:
+        app = client.apps.get(resource_group, service, name)
+    except Exception:
+        # ignore
+        return
+    if app:
+        raise CLIError('App {} already exist.'.format(app.id))
```
Comment (zhoxing-ms) [inline src/spring-cloud/azext_spring_cloud/app.py:366]: Same as above

File: src/spring-cloud/azext_spring_cloud/_app_validator.py [hunk 6]
```diff
@@ -8,8 +8,8 @@
 from azure.cli.core.azclierror import InvalidArgumentValueError
 from msrestazure.azure_exceptions import CloudError
 from azure.core.exceptions import (ResourceNotFoundError)
-from ._client_factory import cf_spring_cloud
 from ._resource_quantity import (validate_cpu as validate_cpu_value, validate_memory as validate_memory_value)
+from ._client_factory import cf_spring_cloud_20220101preview
```
Comment (zhoxing-ms) [inline src/spring-cloud/azext_spring_cloud/_app_validator.py:12]: May I ask why did you add the api-version number to the name of `_client_factory` ?

File: src/spring-cloud/azext_spring_cloud/_buildservices_factory.py (hunk 2 above)
Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/_buildservices_factory.py]: Updated

File: src/spring-cloud/azext_spring_cloud/_buildservices_factory.py (hunk 3 above)
Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/_buildservices_factory.py]: Updated

File: src/spring-cloud/azext_spring_cloud/_buildservices_factory.py (hunk 4 above)
Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/_buildservices_factory.py]: Updated

File: src/spring-cloud/azext_spring_cloud/app.py (hunk 5 above)
Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/app.py]: Updated

File: src/spring-cloud/azext_spring_cloud/_app_validator.py (hunk 6 above)
Comment (yuwzho) [reply src/spring-cloud/azext_spring_cloud/_app_validator.py:12]: Here follows the current implementation about client factory. The default client factory uses the latest stable API version. But some features in previewed is only available in preview API version. Ideally, we, service team will flush all preview features in a new stable API version, at that time, we can remove the `factory_preview_api_version` and replace them as the new default client factory.

Comment (zhoxing-ms) [reply src/spring-cloud/azext_spring_cloud/_app_validator.py:12]: Got it, thanks

**Review by zhoxing-ms — APPROVED** [review]

**Comment by yuwzho** [comment]
#4294
